docs/subnets/hello-world-precompile-tutorial.md

```diff
@@ -0,0 +1,946 @@
+
+When a precompile function is called, it still goes through the `CALL` function in the EVM. However, it works a little differently. The EVM checks if the address is a precompile address from the mapping list and if so redirects to the precompile function.
+
+```go
+  if p := precompiles[addr]; p != nil {
+    return RunPrecompiledContract(p, input, contract)
+  }
+```
+
+The EVM then performs the function and subtracts the `RequiredGas`.
+
+Precompiles provide complex library functions that are commonly used in smart contracts and do not use EVM opcodes which makes execution faster and gas costs lower.
+
+### Stateful Precompiled Contracts
+
+A stateful precompile builds on a precompile in that it adds state access. Stateful precompiles are not available in the default EVM, and are specific to Avalanche EVMs such as [Coreth](https://github.com/ava-labs/coreth) and [Subnet-EVM](https://github.com/ava-labs/subnet-evm). It allows us to add even more functionality and customization to the Subnet-EVM!
+
+A stateful precompile follows this interface.
+
+```go
+// StatefulPrecompiledContract is the interface for executing a precompiled contract
+type StatefulPrecompiledContract interface {
+	// Run executes the precompiled contract.
+	Run(accessibleState PrecompileAccessibleState,
+	caller common.Address,
+	addr  common.Address,
+	input []byte,
+	suppliedGas uint64,
+	readOnly bool)
+	(ret []byte, remainingGas uint64, err error)
+}
+```
+
+Notice the most important difference between the stateful precompile and precompile interface. We now inject state access to the `Run` function. Precompiles only took in a single byte slice as input. However, stateful precompile functions have complete access to the Subnet-EVM state, and can be used to implement a much wider range of functionalities.
+
+With state access, we can modify balances, read/write the storage of other contracts, and could even hook into external storage outside of the bounds of the EVM’s merkle trie (note: this would come with repercussions for state sync since part of the state would be moved off of the merkle trie). We can now write custom logic to make our own EVM.
+
+### Assumption of Knowledge
+
+We assume that the user has knowledge of git, golang, and javascript. It is not necessary to have knowledge of the EVM, but it is extremely helpful. Here are some resources to get started.
```

Comment:
```suggestion
We assume that the user has knowledge of git, golang, and javascript. The user should also be deeply familiar with the EVM since adding a stateful precompile is effectively modifying the EVM and requires an understanding of its invariants to do so correctly.
```